File: tests/integration/TestClipping.cpp
/* Copyright (c) 2015-2022, EPFL/Blue Brain Project
 * All rights reserved. Do not distribute without permission.
 * Responsible Author: Daniel.Nachbaur@epfl.ch
 *
 * This file is part of Brayns <https://github.com/BlueBrain/Brayns>
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License version 3.0 as published
 * by the Free Software Foundation.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include <brayns/Brayns.h>
#include <brayns/engine/camera/projections/Orthographic.h>
#include <brayns/engine/camera/projections/Perspective.h>
#include <brayns/engine/components/ClipperViews.h>
#include <brayns/engine/components/Geometries.h>
#include <brayns/engine/core/Engine.h>
#include <brayns/engine/geometry/types/Plane.h>
#include <brayns/engine/light/types/AmbientLight.h>
#include <brayns/engine/light/types/DirectionalLight.h>
#include <brayns/engine/scene/ModelsOperations.h>
#include <brayns/engine/systems/ClipperInitSystem.h>

#include <tests/paths.h>

#include "helpers/BraynsTestUtils.h"
#include "helpers/ImageValidator.h"

#include <doctest/doctest.h>

namespace
{
struct ZParallelSliceManager
{
    static void generate(brayns::Brayns &brayns)
    {
        auto &engine = brayns.getEngine();
        auto &scene = engine.getScene();
        const auto &bounds = scene.getBounds();
        const auto center = bounds.center();
        const auto dimensions = bounds.dimensions();
        const auto sliceThickness = dimensions.z * 0.1f;

        // Clip every point that dot(planeA, point) + planeADistance < 0.0
        const auto planeA = brayns::Vector3f(0.f, 0.f, 1.f);
        const auto planeADistance = center.z - sliceThickness * 0.5f;

        // Clip every point that dot(planeB, point) + planeBDistance < 0.0
        const auto planeB = brayns::Vector3f(0.f, 0.f, -1.f);
        const auto planeBDistance = -center.z - sliceThickness * 0.5f;

        auto model = std::make_shared<brayns::Model>("");

        auto planes = std::vector<brayns::Plane>{
            brayns::Plane{{planeA, planeADistance}},
            brayns::Plane{{planeB, planeBDistance}}};
        auto &components = model->getComponents();
        auto &clippers = components.add<brayns::Geometries>();
        clippers.elements.emplace_back(std::move(planes));

        auto &systems = model->getSystems();
        systems.setInitSystem<brayns::ClipperInitSystem>();

        auto &models = scene.getModels();
        models.add(std::move(model));
    }

    static void clear(brayns::Brayns &brayns)
    {
        auto &engine = brayns.getEngine();
        auto &scene = engine.getScene();
        auto &models = scene.getModels();
        brayns::ModelsOperations::removeClippers(models);
    }
};
} // namespace

void testClipping(bool orthographic)
{
    const auto original = orthographic ? "testClipBaseOrtho.png" : "testClipBasePerspective.png";
    const auto clipped = orthographic ? "testClipClippedOrtho.png" : "testClipClippedPerspective.png";

    brayns::Brayns brayns;

    BraynsTestUtils::setRenderResolution(brayns, 300, 300);
    BraynsTestUtils::addModel(brayns, BRAYNS_TESTDATA_MODEL_PLY_PATH);

    if (orthographic)
    {
        BraynsTestUtils::adjustOrthographicView(brayns);
    }
    else
    {
        BraynsTestUtils::adjustPerspectiveView(brayns);
    }

    BraynsTestUtils::addLight(brayns, brayns::Light(brayns::DirectionalLight()));
    BraynsTestUtils::addLight(brayns, brayns::Light(brayns::AmbientLight{0.05f}));

    auto &engine = brayns.getEngine();
    brayns.commitAndRender();
    CHECK(ImageValidator::validate(engine, original));

    ZParallelSliceManager::generate(brayns);
    brayns.commitAndRender();
    CHECK(ImageValidator::validate(engine, clipped));
    ZParallelSliceManager::clear(brayns);

    brayns.commitAndRender();
    CHECK(ImageValidator::validate(engine, original));

    ZParallelSliceManager::generate(brayns);
    brayns.commitAndRender();
    CHECK(ImageValidator::validate(engine, clipped));
    ZParallelSliceManager::clear(brayns);
}

TEST_CASE("Perspective clipping")
{
    testClipping(false);
}

TEST_CASE("Orthographic clipping")
{
    testClipping(true);
}

File: python/tests/core/test_clipping.py
# Copyright (c) 2015-2023 EPFL/Blue Brain Project
# All rights reserved. Do not distribute without permission.
#
# Responsible Author: adrien.fleury@epfl.ch
#
# This file is part of Brayns <https://github.com/BlueBrain/Brayns>
#
# This library is free software; you can redistribute it and/or modify it under
# the terms of the GNU Lesser General Public License version 3.0 as published
# by the Free Software Foundation.
#
# This library is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
# details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this library; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

import unittest

import brayns
from tests.mock_instance import MockInstance
from tests.mock_messages import mock_model, mock_model_message


class TestAddClippingGeometries(unittest.TestCase):
    def test_add_clipping_geometry(self) -> None:
        instance = MockInstance(mock_model_message())
        equation = brayns.PlaneEquation(1, 2, 3, 4)
        plane = brayns.ClipPlane(equation)
        test = brayns.add_clipping_geometry(instance, plane)
        self.assertEqual(test, mock_model())
        self.assertEqual(instance.method, brayns.ClipPlane.method)
        self.assertEqual(instance.params, plane.get_properties())

    def test_add_clipping_geometries(self) -> None:
        instance = MockInstance(mock_model_message())
        geometry = brayns.ClippingBox(brayns.Vector3.zero, brayns.Vector3.one)
        test = brayns.add_clipping_geometries(instance, [geometry])
        self.assertEqual(test, mock_model())
        self.assertEqual(instance.method, brayns.ClippingBox.method)
        self.assertEqual(instance.params, [geometry.get_properties()])

    def test_clear_clipping_geometries(self) -> None:
        instance = MockInstance()
        brayns.clear_clipping_geometries(instance)
        self.assertEqual(instance.method, "clear-clip-planes")
        self.assertIsNone(instance.params)

    def test_clip_plane(self) -> None:
        self.assertEqual(brayns.ClipPlane.method, "add-clip-plane")
        equation = brayns.PlaneEquation(1, 2, 3, 4)
        plane = brayns.ClipPlane(equation)
        test = plane.get_properties()
        ref = {"coefficients": [1, 2, 3, 4]}
        self.assertEqual(test, ref)

    def test_clipping_bounded_planes(self) -> None:
        method = brayns.ClippingBoundedPlane.method
        self.assertEqual(method, "add-clipping-bounded-planes")
        equation = brayns.PlaneEquation(1, 2, 3, 4)
        bounds = brayns.Bounds(brayns.Vector3.zero, brayns.Vector3.one)
        plane = brayns.ClippingBoundedPlane(equation, bounds)
        test = plane.get_properties()
        ref = {
            "coefficients": [1, 2, 3, 4],
            "bounds": {
                "min": [0, 0, 0],
                "max": [1, 1, 1],
            },
        }
        self.assertEqual(test, ref)

    def test_clipping_boxes(self) -> None:
        self.assertEqual(brayns.ClippingBox.method, "add-clipping-boxes")
        box = brayns.ClippingBox(brayns.Vector3.zero, brayns.Vector3.one)
        test = box.get_properties()
        ref = {
            "min": [0, 0, 0],
            "max": [1, 1, 1],
        }
        self.assertEqual(test, ref)

    def test_clipping_capsules(self) -> None:
        self.assertEqual(brayns.ClippingCapsule.method, "add-clipping-capsules")
        start = brayns.Vector3.zero
        end = brayns.Vector3.one
        radius = 1
        capsule = brayns.ClippingCapsule(start, radius, end, radius)
        test = capsule.get_properties()
        ref = {
            "p0": [0, 0, 0],
            "r0": 1,
            "p1": [1, 1, 1],
            "r1": 1,
        }
        self.assertEqual(test, ref)

    def test_clipping_planes(self) -> None:
        self.assertEqual(brayns.ClippingPlane.method, "add-clipping-planes")
        equation = brayns.PlaneEquation(1, 2, 3, 4)
        plane = brayns.ClippingPlane(equation)
        test = plane.get_properties()
        ref = {"coefficients": [1, 2, 3, 4]}
        self.assertEqual(test, ref)

    def test_clipping_spheres(self) -> None:
        self.assertEqual(brayns.ClippingSphere.method, "add-clipping-spheres")
        sphere = brayns.ClippingSphere(1)
        test = sphere.get_properties()
        ref = {"center": [0, 0, 0], "radius": 1}
        self.assertEqual(test, ref)
